### ISSUE TITLE
(string[])Request.Headers["NonExistentHeader"] is null while (IEnumerable<string>)Request.Headers["NonExistentHeader"] is not

### ISSUE THREAD

**Issue by hrumhurum** [dotnet/runtime#76903]

**Comment by adityamandaleeka** [comment]
@maryamariyan

**Comment by tarekgh** [comment]
@hrumhurum is this a blocking issue for you or you can work around this for now and we fix it in future releases?

**Comment by eerhardt** [comment]
I'm not sure this suggestion is inline with the original design of `StringValues`. Reading https://github.com/aspnet/Announcements/issues/60, it says:

> The value is implicitly convertable to and from string and string[]

Changing the implicit conversion to `string[]` is going to lose the "nullness" of it. Take the following code as an example:

```C#
string? a = null;
StringValues values = a;

string? b = values;
Console.WriteLine(b is null); // prints 'true'

string?[]? aArray = null;
values = aArray;

string?[]? bArray = values;
Console.WriteLine(bArray is null); // prints 'true' today, but with this change will print 'false'
```

This seems pretty inconsistent to be implicitly convertible to and from a Type, but lose information when you do.

**Comment by hrumhurum** [comment]
@eerhardt, the HTTP spec has no notion of `null` headers, so trying to shove it there is quite meaningless IMHO.

It also means that the conversion from `(string[]?)null` to `StringValues` should be disallowed:

```c#
string[]? a = null;
StringValues values = a; // nullability warning during compilation (but no error in runtime for compatibility reasons)
```

If anybody ever needs the notion of `(string[]?)null` with `StringValues`, it can be achieved in a more natural and mathematically sound way:

```c#
string[]? a = null;
StringValues? values = a;

Console.WriteLine(values is null); // prints 'true'
```

On a general note, I think that the original idea of having a `(string[]?)null` in `StringValues` might well be just to enable the following approach within the header collection:

```c#
Request.Headers["SomeHeader"] = "example"; // sets "SomeHeader" to "example" value

string[]? value = Request.Headers["SomeHeader"]; // retrieves the value of "SomeHeader"
if (value != null)
{
    foreach (var i in value)
        // ...
}
```

The suggested change does not violate that original idea, actually it even improves it making it less prone to errors. Consider the following code (after the change):

```c#
Request.Headers["SomeHeader"] = "example"; // sets "SomeHeader" to "example" value

string[] value = Request.Headers["SomeHeader"]; // retrieves the value of "SomeHeader"
foreach (var i in value)
    // ...
```

**Comment by hrumhurum** [comment]
@tarekgh, this is not a blocking issue in any way as it can be workarounded.

**Comment by eerhardt** [comment]
> the HTTP spec has no notion of null headers

Note that `StringValues` is a primitive type that is usable outside of HTTP headers.

> string value = string.Join("\n", Request.Headers["NonExistentHeader"]);

From your original issue report, you get a nullability warning on this line:

CS8604 - Possible null reference argument for parameter 'value' in 'string string.Join(string? separator, params string?[] value)'.

So it shouldn't be surprising that you need to handle the case that `Request.Headers["NonExistentHeader"]` turns into a `null` string array.

At one point in time there was a proposal to add `IsNull` to StringValues: https://github.com/dotnet/extensions/pull/323. But that was reverted, since it broke the build.

cc @Tratcher

**Comment by davit-asryan-sveasolar** [comment]
> I'm not sure this suggestion is inline with the original design of `StringValues`. Reading https://github.com/aspnet/Announcements/issues/60, it says:
> 
> 
> 
> > The value is implicitly convertable to and from string and string[]
> 
> 
> 
> Changing the implicit conversion to `string[]` is going to lose the "nullness" of it. Take the following code as an example:
> 
> 
> 
> ```C#
> 
> string? a = null;
> 
> StringValues values = a;
> 
> 
> 
> string? b = values;
> 
> Console.WriteLine(b is null); // prints 'true'
> 
> 
> 
> string?[]? aArray = null;
> 
> values = aArray;
> 
> 
> 
> string?[]? bArray = values;
> 
> Console.WriteLine(bArray is null); // prints 'true' today, but with this change will print 'false'
> 
> ```
> 
> 
> 
> This seems pretty inconsistent to be implicitly convertible to and from a Type, but lose information when you do.
> 
> 
> 
> To back up this point, see https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/operator-overloads
> 
> 
> 
> > ❌ DO NOT provide an implicit conversion operator if the conversion is potentially lossy.
> 
> >
> 
> > For example, there should not be an implicit conversion from Double to Int32 because Double has a wider range than Int32. An explicit conversion operator can be provided even if the conversion is potentially lossy.

I agree that lossy implicit conversion is bad.

As I see it the problem here is implicit conversions of string?[]? **to** and from StringValues. It doesn't make sense to me. My suggestion would be to only have implicit conversions for string?[]. And for null values it only makes sense to have null in case of implicit conversion from string?.

**Comment by halter73** [comment]
I agree we shouldn't change `StringValues`. I'm going to move this back to the aspnetcore repo and remove the "help wanted" label until we agree on exactly what we want. We can consider using `new StringValues(Array.Empty<string>())` instead of `new StringValues(null)` since there isn't much of a reason to distinguish between 0 and null header values.
